### PR TITLE
Serializer for EventEnvelope and Offset

### PR DESCRIPTION
**Pull request by patriknw**
Need this for https://github.com/akka/akka-persistence-r2dbc/pull/168

### CODE REVIEW

File: akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+   * The offset is converted from its string representation to its real type.
+   */
+  private def fromStorageRepresentation(offsetStr: String, manifest: String): Offset = {
+    manifest match {
+      case TimestampOffsetManifest     => timestampOffsetFromStorageRepresentation(offsetStr)
+      case SequenceOffsetManifest      => Offset.sequence(offsetStr.toLong)
+      case TimeBasedUUIDOffsetManifest => Offset.timeBasedUUID(UUID.fromString(offsetStr))
+      case NoOffsetManifest            => NoOffset
+      case _ =>
+        val parts = manifest.split(':')
```
Comment (patriknw) [inline akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala:123]: this is in the same as the String format used in Akka Projections

**Comment by patriknw** [comment]
Downloading adopt jdks

File: akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala (hunk 1 above)
Comment (johanandren) [reply akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala:123]: make it a named constant for extra clarity

File: akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        val bytes = serializer.toBinary(obj)
+        val offsetStr = Base64.getEncoder.encodeToString(bytes)
+        (offsetStr, s"$serializerId:$serializerManifest")
+    }
+  }
+
+  private def timestampOffsetFromStorageRepresentation(str: String): TimestampOffset = {
+    try {
+      val parts = str.split(timestampOffsetSeparator)
+      if (parts.length == 4) {
```
Comment (johanandren) [inline akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala:166]: Not important, and subjective, but could perhaps be cleaner as a patmatch
```scala
parts match {
  case Array(ts, rts, seenPid, seenSeqnr) =>...
  case Array(ts) => 
  case Array(ts, rts) => 
  case lotsaparts =>
```

**Review by johanandren — APPROVED** [review]
LGTM

File: akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+  private final val EventEnvelopeManifest = "a"
+
+  private final val SequenceOffsetManifest = "SEQ"
+  private final val TimeBasedUUIDOffsetManifest = "TBU"
+  private final val TimestampOffsetManifest = "TSO"
+  private final val NoOffsetManifest = "NO"
+
+  // persistenceId and timestamp must not contain this separator char
+  private val timestampOffsetSeparator = ';'
```
[the remark this answers is not in the document]
Comment (octonato) [reply akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala:49]: Or we could relax one this and let it fail with a clear message. 
Then we make this configurable allowing to pick a different separator? Am I overthinking this? 🤔

File: akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala (hunk 2 above)
Comment (octonato) [reply akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala:166]: I find this important. 😄

**Review by octonato — APPROVED** [review]
LGTM, with a remark, but non-blocking.

**Comment by patriknw** [comment]
@johanandren @octonato pls take another look, I incorporated your feedback in https://github.com/akka/akka/pull/30970/commits/70121771286853f8505f5250a7f29f051632a88f

File: akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala:50]: I added a check in the toStorageRepresentation

**Review by johanandren — APPROVED** [review]

**Review by octonato — APPROVED** [review]
LGTM

**Pull request merged**